### PR TITLE
Update to python 3.10

### PR DESCRIPTION
**Pull request by AstroFloof**

### CODE REVIEW

File: bot/bot.py [hunk 1]
```diff
@@ -9,17 +9,22 @@
...
+load_dotenv()
 
 
 ENVVARS = ["FRED_IP", "FRED_PORT", "FRED_TOKEN", "DIALOGFLOW_AUTH",
            "FRED_SQL_DB", "FRED_SQL_USER", "FRED_SQL_PASSWORD",
            "FRED_SQL_HOST", "FRED_SQL_PORT"]
 
 for var in ENVVARS:
-    assert (os.environ.get(var)), f"The ENV variable '{var}' isn't set"
+    if not (os.environ.get(var)):
```
Comment (Feyko) [inline bot/bot.py:26]: No need for the parentheses now

File: bot/bot.py [hunk 2]
```diff
@@ -42,7 +47,8 @@ def __init__(self, *args, **kwargs):
         self.setup_cogs()
         self.version = "2.17.5"
 
-        self.loop = asyncio.get_event_loop()
+        self.loop = asyncio.new_event_loop()
```
Comment (Feyko) [inline bot/bot.py:50]: why

File: bot/bot.py [hunk 3]
```diff
@@ -141,23 +147,23 @@ async def githook_send(self, data):
...
             await user.create_dm()
         try:
             if not embed:
                 embed = createembed.DM(content)
                 content = None
-            return await user.dm_channel.send(content=content, embed=embed, file=file)
+            return await user.dm_channel.send(content=content, embed=embed, file=file, **kwargs)
         except Exception as e:
-            print(e)
+            logging.error(f"DMs: {e}")
```
Comment (Feyko) [inline bot/bot.py:166]: Add that we failed to send a DM

File: bot/cogs/crashes.py [hunk 4]
```diff
@@ -189,113 +189,119 @@ async def check_for_outdated_mods(self, mod_list: list):
...
-                        if not sml_version and fg_sml_version:
-                            sml_version = fg_sml_version
-                        if not game_version and fg_game_version:
-                            game_version = fg_game_version
-                        if not path and fg_path:
-                            path = fg_path
-                        if not launcher_id and fg_launcher_id:
-                            launcher_id = fg_launcher_id
-                        if not commandline and fg_commandline:
-                            commandline = fg_commandline
```
Comment (Feyko) [inline bot/cogs/crashes.py:246]: Why the huge diff? What did you actually change?

**Review by Feyko — COMMENTED** [review]
Since you touched the commit formatting for the githook, please test it to make sure it looks the same (or at least doesn't look whack)

File: bot/bot.py (hunk 2 above)
Comment (AstroFloof) [reply bot/bot.py:50]: because there isn’t one to get?

File: bot/cogs/crashes.py (hunk 4 above)
Comment (AstroFloof) [reply bot/cogs/crashes.py:246]: it’s inside a match statement and git decided to be an asshat with indentation commits

File: bot/bot.py (hunk 2 above)
Comment (Feyko) [reply bot/bot.py:50]: asyncio has a default loop

File: bot/cogs/crashes.py (hunk 4 above)
Comment (Feyko) [reply bot/cogs/crashes.py:246]: fUn

File: bot/bot.py (hunk 2 above)
Comment (AstroFloof) [reply bot/bot.py:50]: as far as i know it wasn't working for me at some point. best to keep what works, as we weren't even referencing the loop before

**Comment by AstroFloof** [comment]
> Since you touched the commit formatting for the githook, please test it to make sure it looks the same (or at least doesn't look whack)

As you already know, I have now done that.
But I like making this stuff official on GitHub so yeah.

File: bot/bot.py [hunk 5]
```diff
@@ -168,11 +168,14 @@ async def send_DM(user, content, embed=None, file=None, **kwargs):
 
     @staticmethod
     async def reply_to_msg(message, content=None, propagate_reply=True, **kwargs):
+        logging.info(f"Replying to a message", extra=common.messagedict(message))
+        # use this line if you're trying to debug discord throwing code 400s
+        # logging.debug(jsonpickle.dumps(dict(content=content, **kwargs), indent=2))
         reference = (message.reference if propagate_reply else None) or message
         if isinstance(reference, discord.MessageReference):
             reference.fail_if_not_exists = False
-        logging.info(f"Replying to a message", extra=helper.messagedict(message))
```
Comment (Feyko) [inline bot/bot.py:168]: Why remove the logging

File: bot/bot.py [hunk 6]
```diff
@@ -168,11 +168,14 @@ async def send_DM(user, content, embed=None, file=None, **kwargs):
...
-        return await message.channel.send(content, reference=reference, **kwargs)
+
+        return await message.reply(content, **kwargs)
```
Comment (Feyko) [inline bot/bot.py:178]: This was intended. Propagates the reference if the calling message is a reply to another message

File: bot/bot.py [hunk 7]
```diff
@@ -218,19 +221,25 @@ async def on_message(self, message):
...
+            if not removed:
+                if message.content.startswith(self.command_prefix):
+                    await self.process_commands(message)
+                else:
+                    reacted = await self.Crashes.process_message(message)
+                    if not reacted:
+                        await self.DialogFlow.process_message(message)
+            logging.info("Finished processing a message", extra=common.messagedict(message))
+        except:
+            logging.error(traceback.format_exc())
```
Comment (Feyko) [inline bot/bot.py:235]: Please add context to the error (what we were trying to do)

File: bot/cogs/crashes.py [hunk 8]
```diff
@@ -24,6 +23,7 @@
 def timeout(seconds=2, error_message=os.strerror(errno.ETIME)):
     def decorator(func):
         def _handle_timeout(signum, frame):
+            logging.error("A regex timed out!")
```
Comment (Feyko) [inline bot/cogs/crashes.py:26]: Add some info about the regex

File: bot/cogs/crashes.py [hunk 9]
```diff
@@ -45,59 +45,63 @@ def timedregexsearch(*args, **kwargs):
     try:
         return re.search(*args, **kwargs)
     except TimeoutError:
-        raise TimeoutError(f"The following regexp timed out: '{args[0]}'")
+        logging.error(TimeoutError(f"The following regexp timed out: '{args[0]}'"))
```
Comment (Feyko) [inline bot/cogs/crashes.py:48]: Does logging.error re-raise the error?

Comment (Feyko) [reply bot/cogs/crashes.py:48]: Also, that's two logs for a regex timeout

**Review by AstroFloof — COMMENTED** [review]
Reviewed at 12:34pm EST on Jan 5, 2022

File: bot/cogs/crashes.py [hunk 10]
```diff
@@ -45,59 +45,63 @@ def timedregexsearch(*args, **kwargs):
...
         if path:
             version_info += f"Path: {path}\n"
-        if launcher_id:
-            version_info += f"Launcher ID: {launcher_id}\n"
-        if commandline:
-            version_info += f"Command Line: {commandline}\n"
+        if launcher:
+            version_info += f"Launcher ID: {launcher}\n"
+        if cli.strip():
+            version_info += f"Command Line: {cli}\n"
```
Comment (Feyko) [inline bot/cogs/crashes.py:99]: I liked the explicit names. Also, cli is command line *interface* :P

Comment (AstroFloof) [reply bot/cogs/crashes.py:99]: think of it as command line input :P

File: bot/bot.py (hunk 5 above)
Comment (AstroFloof) [reply bot/bot.py:168]: I moved it to the top of the function

File: bot/cogs/commands.py [hunk 11]
```diff
@@ -34,25 +33,29 @@ def __init__(self, bot):
     @commands.Cog.listener()
     async def on_command_error(self, ctx, error):
         # We get an error about commands being found when using "runtime" commands, so we have to ignore that
+        self.bot.logger.info("Handling command error")
```
Comment (Feyko) [inline bot/cogs/commands.py:36]: This'll get logged with every message that isn't a command. Not sure if we want that. Could replace it by explicitly stating that the message wasn't a command in the first if below this line (if this log doesn't already exist in commands.py)

File: bot/cogs/crashes.py [hunk 12]
```diff
@@ -23,7 +23,6 @@
 def timeout(seconds=2, error_message=os.strerror(errno.ETIME)):
     def decorator(func):
         def _handle_timeout(signum, frame):
-            logging.error("A regex timed out!")
```
Comment (Feyko) [inline bot/cogs/crashes.py]: Please move this to timedregexsearch

Comment (AstroFloof) [reply bot/cogs/crashes.py]: I reverted my changes here because all the previous timed regex logging is sufficint

**Pull request merged**